### PR TITLE
Bp 3311 add support for woo commerce block based checkout

### PR DESCRIPTION
**Pull request by vegimcarkaxhija**

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -108,9 +108,98 @@ function buckaroo_payment_setup_scripts()
...
+
+/**
+ * Check if payment gateway is ours
+ *
+ * @param string $name
+ *
+ * @return boolean
+ */
+function isBuckarooPayment(string $name): bool {
+	return str_starts_with( $name, 'buckaroo' );
```
Comment (harli91) [inline index.php:147]: what about php 7.4

File: index.php [hunk 2]
```diff
@@ -108,9 +108,98 @@ function buckaroo_payment_setup_scripts()
...
+function get_active_issuer_code()
+{
+    return BuckarooPayByBank::getActiveIssuerCode();
+}
+
+function get_type() {
+	return (new WC_Gateway_Buckaroo_Afterpay())->type;
+}
+
+function get_credtCard_is_secure() {
```
Comment (harli91) [inline index.php:133]: these function pollute the global namespace

File: assets/js/checkout/creditcard-call-encryption.js [hunk 3]
```diff
@@ -76,6 +95,7 @@ class BuckarooValidateCreditCards {
             cardholder,
             function(encryptedCardData) {
                 parent.find(".encryptedCardData").val(encryptedCardData);
+                jQuery(document).trigger("encryptedDataChanged", [encryptedCardData]);
```
Comment (harli91) [inline assets/js/checkout/creditcard-call-encryption.js:98]: why aren't you running the encrypt function directly in react?

File: gateway-buckaroo.php [hunk 4]
```diff
@@ -507,6 +518,7 @@ public function request($key)
         if (!isset($_POST[$key])) {
            return;
         }
+//        var_dump($_POST);
```
Comment (harli91) [inline gateway-buckaroo.php:521]: remove comment

File: index.php [hunk 5]
```diff
@@ -108,9 +108,98 @@ function buckaroo_payment_setup_scripts()
             'in3_v3' => WC_Gateway_Buckaroo_In3::IN3_V3_TITLE,
         )
     );
+
+	wp_enqueue_script('my-block-script', 'assets/js/dist/blocks.js', array('wp-blocks', 'wp-element'));
```
Comment (harli91) [inline index.php:112]: maybe name this less generic

File: assets/js/checkout/creditcard-call-encryption.js (hunk 3 above)
Comment (vegimcarkaxhija) [reply assets/js/checkout/creditcard-call-encryption.js:98]: this file is used also for normal checkout without React, than I should create two files that have the same function

File: assets/js/blocks/components/BuckarooLabel.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+export const BuckarooLabel = ({ image_path, title }) => {
+    return (
+        <div className='buckaroo_method_block'>
+            {title}
+            <img src={image_path} alt='Payment Method' style={{ float: 'right' }} />
```
Comment (harli91) [inline assets/js/blocks/components/BuckarooLabel.js:5]: set the alt to the payment method name

File: assets/js/blocks/gateways/buckaroo_afterpay.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        onBirthdateChange,
+        onCocInput,
+        onCompanyInput,
+        onAccountName,
+        onCocRegistrationChange,
+        onAdditionalCheckboxChange,
+    }= callbacks;
+
+    const paymentMethod = 'buckaroo-afterpay';
+    const [isAdditionalCheckboxChecked, setIsAdditionalCheckboxChecked] = useState(false);
```
Comment (harli91) [inline assets/js/blocks/gateways/buckaroo_afterpay.js:30]: why do we need this state?

File: assets/js/checkout/checkout.js [hunk 8]
```diff
@@ -34,19 +34,21 @@ class BuckarooCheckout {
...
 
       if (jQuery("#billing_company").length) {
-        jQuery("#buckaroo-afterpay-CompanyName").val(
+        jQuery("#buckaroo-afterpay-company-name").val(
           jQuery("#billing_company").val()
         );
       }
 
-      birthdate.prop("disabled", !b2bActive);
+      buckaroo_partial_birth_field.toggle(!b2bActive);
```
Comment (harli91) [inline assets/js/checkout/checkout.js:50]: what is this field, i cannot se this field added in the php templated

File: assets/js/checkout/creditcard-call-encryption.js (hunk 3 above)
Comment (harli91) [reply assets/js/checkout/creditcard-call-encryption.js:98]: I think we better separate the code from the normal checkout and the react checkout, if we remove the normal checkout functionality it should not brake the react functionality

File: library/common.php [hunk 9]
```diff
@@ -1157,26 +1157,54 @@ function buckaroo_request_sanitized_json($key)
...
-            $genders = ['male' => 'male', 'female' => 'female'];
-            break;
-        case 'buckaroo-klarnapii' :
-            $genders = ['male' => 'male', 'female' => 'female'];
-            break;
-        default :
-            $genders = ['male' => 'male', 'female' => 'female', 'they'=> 'unknown', 'unknown' => 'unknown'];
-            break;
-    }
+function getAllGendersForPaymentMethods(): array {
```
Comment (harli91) [inline library/common.php:1160]: move this into a class, use proper constants

**Review by harli91 — CHANGES_REQUESTED** [review]

File: assets/js/blocks/gateways/buckaroo_afterpaynew.js [hunk 10]
```diff
@@ -1,71 +1,99 @@
...
+        onCocInput
+    }= callbacks;
 
-                {(billingData.country === 'BE' || billingData.country === 'NL') && (
-                    <div>
-                        <BirthDayField paymentMethod={paymentMethod} onBirthdateChange={onBirthdateChange} />
-                    </div>
-                )}
+    const paymentMethod = 'buckaroo-afterpaynew';
+    const [isTermsAccepted, setIsTermsAccepted] = useState(false);
```
Comment (harli91) [inline assets/js/blocks/gateways/buckaroo_afterpaynew.js]: why do you need this state?

File: assets/js/blocks/gateways/buckaroo_afterpaynew.js [hunk 11]
```diff
@@ -1,71 +1,99 @@
...
 
-                {billingData.country === 'NL' && customer_type !== 'b2c' && (
-                    <p className="form-row form-row-wide validate-required">
-                        <label htmlFor="buckaroo-afterpaynew-coc">
-                            {__('CoC-number:', 'wc-buckaroo-bpe-gateway')}
-                            <span className="required">*</span>
-                        </label>
+    const handleChange = (field, value) => {
+        switch (field) {
+            case 'terms':
```
Comment (harli91) [inline assets/js/blocks/gateways/buckaroo_afterpaynew.js]: why do you need this swich?

File: index.php [hunk 12]
```diff
@@ -138,13 +126,22 @@ function get_credtCard_is_secure() {
...
+}
+
+/**
+ * Check if a string starts with a specific prefix
+ *
+ * @param string $haystack
+ * @param string $needle
+ * @return bool
+ */
+function starts_with(string $haystack, string $needle): bool {
```
Comment (harli91) [inline index.php:145]: don't pollute the global namespace

File: assets/js/blocks/partials/buckaroo_gender.js [hunk 13]
```diff
@@ -1,11 +1,21 @@
...
 
 const GenderDropdown = ({ paymentMethod, genders, onSelectGender }) => {
+    const translateGender = (key) => {
+        const translations = {
+            'male': 'He/him',
+            'female': 'She/her',
+            'they': 'They/them',
+            'unknown': 'I prefer not to say'
+        };
+        return translations[key] || key;
```
Comment (harli91) [inline assets/js/blocks/partials/buckaroo_gender.js:13]: maybe set the key to capitalize when you return it

File: gateway-buckaroo-billink.php [hunk 14]
```diff
@@ -48,10 +48,9 @@ public function validate_fields()
              ) {
                 wc_add_notice(__("Please enter correct birth date", 'wc-buckaroo-bpe-gateway'), 'error');
             }
-
-	        if (!in_array($this->request('buckaroo-billink-gender'), ["0", "1", "2", "9"])) {
-		        wc_add_notice(__("Unknown gender", 'wc-buckaroo-bpe-gateway'), 'error');
-	        }
+            if(!in_array($this->request('buckaroo-billink-gender'), ["Male","Female","Unknown"])) {
```
Comment (harli91) [inline gateway-buckaroo-billink.php]: what about the value further down the line, i don't see it changed

File: assets/js/blocks/partials/buckaroo_terms_and_condition.js [hunk 15]
```diff
@@ -32,7 +32,7 @@ const TermsAndConditionsCheckbox = ({paymentMethod, onCheckboxChange, billingDat
     }
     const handleCheckboxChange = () => {
         setIsChecked(!isChecked);
-        onCheckboxChange(!isChecked ? 'On' : 'Off');
+        onCheckboxChange(!isChecked ? 'On' : null);
```
Comment (harli91) [inline assets/js/blocks/partials/buckaroo_terms_and_condition.js:37]: why do you need to set it here as 'On'?

File: assets/js/blocks/index.js [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+    const [cardCVC, setCardCVCChange] = useState('');
+    const [email, setEmail] = useState('');
+    const [creditCard, setCreditCard] = useState('');
+    const [PaymentComponent, setPaymentComponent] = useState(null);
+    const [cocNumber, setCocNumber] = useState('');
+    const [cocRegistration, setCocRegistrationChange] = useState('');
+    const [encryptedData, setEncryptedDataChange] = useState('');
+    const [identificationNumber, setIdentificationNumber] = useState('');
+    const [companyName, setCompanyName] = useState('');
+    const [additionalCheckboxChange, setAdditionalCheckboxChange] = useState('');
```
Comment (harli91) [inline assets/js/blocks/index.js:51]: split these states into providers for each payment method, a massive global state will be hard to maintain and debug

File: assets/js/blocks/components/BuckarooClientSideEncryption.js [hunk 17]
```diff
@@ -0,0 +1,23 @@
+import BuckarooClientSideEncryption from '../../checkout/creditcard-encryption-sdk';
+
+const encryptCardData = (cardDetails, onEncryptedDataChange) => {
```
Comment (harli91) [inline assets/js/blocks/components/BuckarooClientSideEncryption.js:3]: why aren't you using a async function here? and return the encrypted result directly?

File: library/config.php [hunk 18]
```diff
@@ -212,5 +218,53 @@ public static function getIdinMode() {
...
+            'male' => self::GENDER_MALE,
+            'female' => self::GENDER_FEMALE
+        ];
+        $allGenders['buckaroo-klarnapay'] = [
+            'male' => self::GENDER_MALE,
+            'female' => self::GENDER_FEMALE
+        ];
+        $allGenders['buckaroo-klarnapii'] = [
+            'male' => self::GENDER_MALE,
+            'female' => self::GENDER_FEMALE
```
Comment (harli91) [inline library/config.php:248]: these have basically the same values, can't you just extract it into a variable?
also why not set `$allGenders = ['buckaroo-klarnakp' => [],'buckaroo-klarnapay' ]` looks a lot cleaner to me

File: assets/js/blocks/components/BuckarooClientSideEncryption.js (hunk 17 above)
Comment (harli91) [reply assets/js/blocks/components/BuckarooClientSideEncryption.js:3]: also this is a function not a component, so its better if we move it in a helper folder or something similar

**Review by harli91 — APPROVED** [review]

**Pull request merged**